azext_iot/tests/iothub/__init__.py

```diff
@@ -94,31 +96,38 @@ def __init__(self, test_scenario, add_data_contributor=True):
                 ).get_output_in_json()
 
                 if add_data_contributor:
-                    account = self.cmd("account show").get_output_in_json()
-                    user = account["user"]
-
-                    if user["name"] is None:
-                        raise Exception("User not found")
-
-                    tries = 0
-                    while tries < MAX_RBAC_ASSIGNMENT_TRIES:
-                        role_assignments = self.get_role_assignments(target_hub["id"], USER_ROLE)
-                        role_assignment_principal_names = [assignment["principalName"] for assignment in role_assignments]
-                        if user["name"] in role_assignment_principal_names:
-                            break
-                        # else assign IoT Hub Data Contributor role to current user and check again
-                        self.cmd(
-                            'role assignment create --assignee "{}" --role "{}" --scope "{}"'.format(
-                                user["name"], USER_ROLE, target_hub["id"]
-                            )
-                        )
-                        sleep(10)
+                    self._add_data_contributor(target_hub)
 
-                    if tries == MAX_RBAC_ASSIGNMENT_TRIES:
-                        raise Exception(
-                            "Reached max ({}) number of tries to assign RBAC role. Please re-run the test later "
-                            "or with more max number of tries.".format(MAX_RBAC_ASSIGNMENT_TRIES)
-                        )
+        else:
+            # when the hub instance is existed, make sure to add data contributor role
+            account = self.cmd("account show").get_output_in_json()
+            user = account["user"]
+
+            target_hub = self.cmd(
+                "iot hub show -n {} -g {}".format(self.entity_name, self.entity_rg)
+            ).get_output_in_json()
+
+            role_assignments = self.cmd(
+                'role assignment list --assignee "{}" --scope "{}" --role "{}"'.format(
+                    user["name"], target_hub["id"], USER_ROLE
+                )
+            ).get_output_in_json()
+
+            if len(role_assignments) == 0:
+                self._add_data_contributor(target_hub)
+
+            # Note down current devices and configs so they can be excluded from tear down when tests are done
+            self.original_device_list.extend(d["deviceId"] for d in self.cmd(
```

Comment:
In a real life test scenario you would want to also make sure that none of the new devices, deployments, etc would share a name with the current devices, deployments, etc 

But this is **very** unlikely since the generated ids for the devices should end up being different enough. So I would not mind if you skip this.

Reply:
good point, how about I do check when generating new device id to make it's not repetitive to the old(existed device) names? Also, this kind of come back to @c-ryan-k's suggestion, should we clean up the devices for existed resource before running these tests?

Reply:
What do you think? First decide if the hub should be cleaned or not: Do you think if you give a hub, it should be cleaned before it is used (will help with weirdness, like lets say you have a config that would affect every device; the test resource is a test resource so it should not contain anything that you want to always keep) or should the current dataplane be kept (maybe you want to test out how that config would react with newly created devices in the test; maybe you are proud of your 10k device hub)?